### PR TITLE
Split diff by block type

### PR DESCRIPTION
**Pull request by KevinMind**
Relates to: https://github.com/mozilla/addons-server/pull/22826
Relates to: https://github.com/mozilla/addons-server/pull/22828
Fixes: mozilla/addons#15147

### Description

Split `generate_diffs` to include diffs for both hard and soft blocks. This is required for both stashing and filter updates.

### Context

This patch is required by both of the linked PRs and to keep them independent we should land this (trivial) change first and rebase the other two.

### Testing

This is a backaward compatible patch so automated testing is fine.

### Checklist

- [X] Add `#ISSUENUM` at the top of your PR to an existing open issue in the mozilla/addons repository.
- [X] Successfully verified the change locally.
- [X] The change is covered by automated tests, or otherwise indicated why doing so is unnecessary/impossible.
- [ ] Add before and after screenshots (Only for changes that impact the UI).
- [ ] Add or update relevant [docs](../docs/) reflecting the changes made.

### CODE REVIEW

File: src/olympia/blocklist/mlbf.py [hunk 1]
```diff
@@ -19,6 +19,15 @@
 log = olympia.core.logger.getLogger('z.amo.blocklist')
 
 
+def diff_lists(
+    previous: List[str], current: List[str]
+) -> Tuple[Set[str], Set[str], int]:
+    extras = set(current) - set(previous)
+    deletes = set(previous) - set(current)
+    changed_count = len(extras) + len(deletes) if len(previous) > 0 else len(current)
```
Comment (eviljeff) [inline src/olympia/blocklist/mlbf.py:27]:
```suggestion
    changed_count = len(extras) + len(deletes)
```
the `if` is redundant?

Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py]: Done.

File: src/olympia/blocklist/mlbf.py [hunk 2]
```diff
@@ -148,16 +166,18 @@ def soft_blocked_items(self) -> List[str]:
...
         # even though we exclude all the version ids in the query there's an
         # edge case where the version string occurs twice for an addon so we
         # ensure not_blocked_items contain no blocked_items or soft_blocked_items.
-        return list(
-            set(not_blocked_items) - set(self.blocked_items + self.soft_blocked_items)
-        )
+        return [
+            item
+            for item in not_blocked_items
+            if item not in self.blocked_items + self.soft_blocked_items
```
Comment (willdurand) [inline src/olympia/blocklist/mlbf.py:179]: It looks like the second branch in this condition is not covered by tests.

With this diff applied, I would expect some tests to break but `test_mlbf.py` passes.
```diff
diff --git a/src/olympia/blocklist/mlbf.py b/src/olympia/blocklist/mlbf.py
index b5e7bbbfad..ed8d658452 100644
--- a/src/olympia/blocklist/mlbf.py
+++ b/src/olympia/blocklist/mlbf.py
@@ -176,7 +176,7 @@ class MLBFDataBaseLoader(BaseMLBFLoader):
         return [
             item
             for item in not_blocked_items
-            if item not in self.blocked_items + self.soft_blocked_items
+            if item not in self.blocked_items # + self.soft_blocked_items
         ]
```

File: src/olympia/blocklist/mlbf.py [hunk 3]
```diff
@@ -148,16 +166,18 @@ def soft_blocked_items(self) -> List[str]:
...
-            )
+            Version.unfiltered.exclude(id__in=all_blocks_ids or ())
+            .order_by('id')
+            .values_list('addon__addonguid__guid', 'version')
         )
         # even though we exclude all the version ids in the query there's an
         # edge case where the version string occurs twice for an addon so we
         # ensure not_blocked_items contain no blocked_items or soft_blocked_items.
-        return list(
-            set(not_blocked_items) - set(self.blocked_items + self.soft_blocked_items)
```
Comment (willdurand) [inline src/olympia/blocklist/mlbf.py:159]: I am a bit nervous about getting rid of the sets because of https://github.com/mozilla/addons/issues/15132 but OTOH `not_blocked_items` should only contain unique entries already.

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 4]
```diff
@@ -141,13 +143,32 @@ def test_load_returns_expected_data(self):
...
+            [
+                (block_version.block.guid, block_version.version.version)
+                for block_version in blocked_versions
+            ]
+        )
+        assert mlbf_data[MLBFDataType.SOFT_BLOCKED] == MLBF.hash_filter_inputs(
+            [
+                (soft_blocked_version.block.guid, soft_blocked_version.version.version)
+                for soft_blocked_version in soft_blocked_versions
+            ]
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:171]: can we assert the `len()` as well? it should 10 for each type, right?

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 5]
```diff
@@ -193,6 +214,47 @@ def test_get_data_from_db(self):
         mlbf_data = MLBFDataBaseLoader(mlbf.storage)
         assert mlbf.data._raw == mlbf_data._raw
 
+    def test_cache_json_is_sorted(self):
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:217]: I don't think this test is quite right because if I apply it to current `master`, it passes.

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 6]
```diff
@@ -300,44 +365,105 @@ def test_diff_block_added_and_removed(self):
...
         with next_mlbf.storage.open(next_mlbf.stash_path, 'r') as f:
             assert json.load(f) == {
                 'blocked': [],
                 'unblocked': MLBF.hash_filter_inputs(
-                    [(block_version.block.guid, block_version.version.version)]
+                    [
+                        (block_version.block.guid, block_version.version.version)
+                        for i, block_version in enumerate(block_versions)
+                        if i % 2 == 0
+                    ]
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:466]: This is very hard to follow what's going on and what is being asserted. I'd suggest you delete one or two `block_version` before generating the next mlbf, and then you use that instead, but without a loop or modulo.

**Review by willdurand — CHANGES_REQUESTED** [review]
This is looking promising. I have some comments.

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 4 above)
Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:171]: The len is asserted via the equality check but I cleaned up the code a bit to make it more obbvious by using a single 10 length loop

File: src/olympia/blocklist/mlbf.py (hunk 2 above)
Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py:179]: I tried reproducing this and both are covered by the duplicate test as we create a hard and soft block from the same reused addon

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 5 above)
Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:217]: If you change the order by in the queries the test fails.

File: src/olympia/blocklist/mlbf.py (hunk 3 above)
Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py:159]: I was looking at this and I'm not even sure how I would test it... I added `distinct()` to the queries but that's kind of moot as we have unique indexes for  all the relevant combinations.

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 5 above)
Comment (willdurand) [reply src/olympia/blocklist/tests/test_mlbf.py:237]: Do the queries on `master` have ORDER BY clauses? The test doesn't capture the "broken" behavior that this PR is supposed to fix. I would expect a failing test on `master` and a passing test with this PR.

**Review by willdurand — APPROVED** [review]
[r+wc](https://forum.bors.tech/t/what-does-r-wc-mean-within-the-mozreview-system/101), thanks

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 5 above)
Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:237]: They do not.. and for me at least it fails on master. It could be that this test is actually flaky on master.. if we don't specify ordering it _could_ be returned in the right order and the fewer number of items in the list the higher the probability would be.

This is from latest master right now.

image

Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:237]: Currently running a 100 loop on it and it is indeed flaky. Sometimes passes.
